Source/DafnyRuntime/DafnyRuntimeRust/src/lib.rs

```
@@ -9,6 +16,346 @@ pub use num::NumCast;
+        for elem in array.iter() {
+            result.push(elem_converter(elem));
+        }
+        Sequence::<T>::new_array_sequence(&Rc::new(result))
+    }
+    
+    fn dafny_map_to_hashmap<K, V, K2, V2>(m: &Rc<Map<K, V>>, converter_k: fn(&K)->K2, converter_v: fn(&V)->V2) -> HashMap<K2, V2>
+      where
+          K: Eq + Clone + Hash, V: Clone,
+          K2: Eq + Hash, V2: Clone
+    {
+        m.to_hashmap_owned(converter_k, converter_v)
+    }
+
+    fn hashmap_to_dafny_map<K2, V2, K, V>(map: &HashMap<K2, V2>, converter_k: fn(&K2)->K, converter_v: fn(&V2)->V)
+        -> Rc<Map<K, V>>
+      where
+        K: Eq + Clone + Hash, V: Clone,
+        K2: Eq + Hash, V2: Clone
+    {
+        let mut result: Vec<(K, V)> = Vec::new();
+        for (k, v) in map.iter() {
+            result.push((converter_k(k), converter_v(v)));
+        }
+        let s = Sequence::<(K, V)>::new_array_sequence(&Rc::new(result));
+        Rc::new(Map {
+            data: s,
+            cache: RefCell::new(None)
+        })
+    }
+
+    // --unicode-chars:true
+    pub mod unicode_chars_true {
+        use crate::Sequence;
+        use std::rc::Rc;
+
+        pub fn string_to_dafny_string(s: &str) -> Rc<Sequence<char>> {
+            Sequence::new_array_sequence_is_string(&Rc::new(s.chars().collect()), true)
+        }
+        pub fn dafny_string_to_string(s: &Rc<Sequence<char>>) -> String {
+            let characters = s.to_array();
+            characters.iter().collect::<String>()
+        }
+    }
+    
+    // --unicode-chars:false
+    pub mod unicode_chars_false {
+        use crate::Sequence;
+        use std::rc::Rc;
+        pub fn string_to_dafny_string(s: &str) -> Rc<Sequence<u16>> {
+            Sequence::new_array_sequence_is_string(&Rc::new(s.encode_utf16().collect()), false)
+        }
+        pub fn dafny_string_to_string(s: &Rc<Sequence<u16>>) -> String {
+            let characters = s.to_array();
+            String::from_utf16_lossy(&characters)
```

Wouldn't it be better to panic on invalid UTF16 instead?

---

Since dafny strings are built only from valid u16 that are UTF-16, this problem will never arise, but I'll do the change in this PR.